### ISSUE TITLE
Exception thrown when flattening JSON contains properties with same name on different levels

### ISSUE THREAD

**Issue by gjvanvuuren** [Cinchoo/ChoETL#179]
Hi,

I'm pleasantly surprised with how powerful this package is, but I'm experiencing the following issue:

I'm trying to flatten a relatively simple JSON into a table structure (CSV, DataTable, whatever). My problem is this: When a duplicate JSON property **key** appears on two different levels, an exception is thrown: _"Can not add property ComponentId to Newtonsoft.Json.Linq.JObject. Property with the same name already exists on object."_

The offending properties below are **jsonPath: "ComponentId"**, and **jsonPath: "AxisData[*].ComponentId"**.

I've tried all the tweaks I could think of. I'd appreciate a suggestion on how to resolve this blocking issue.

Here's a reproduceable example: 

```
var json = @"{
""Message"": ""MsgName"",
""TimestampLocal"": ""2022-02-02T12:06:18.3400276+11:00"",
""TimestampGlobal"": ""2022-02-02T00:58:46.1036028Z"",
""TargetComponentId"": ""compid_a"",
""AxisData"": [
  {
    ""MaxTemperature"": 33.1,
    ""PositionFrom"": 2660.0,
    ""PositionTo"": 1311.0,
    ""Name"": ""Xaxis"",
    ""ComponentId"": ""Xaxis_compid"",
    ""MaxCurrent"": 7692.0
  },
  {
    ""MaxTemperature"": 31.9,
    ""PositionFrom"": 2145.0,
    ""PositionTo"": 254.0,
    ""Name"": ""Zaxis"",
    ""ComponentId"": ""Zaxis_compid"",
    ""MaxCurrent"": 6566.0
  },
  {
    ""PositionFrom"": 90.0,
    ""PositionTo"": -90.0,
    ""Name"": ""Caxis"",
    ""ComponentId"": ""Caxis_compid"",
    ""MaxCurrent"": 4432.0
  }
],
""ActionId"": ""87990"",
""ComponentId"": ""compid_b"",
""Duration"": 0,
""TransactionId"": ""0b10e099-69b8-4a8e-b704-d087ef0d6915""
}";

StringBuilder csv = new StringBuilder();

var conf = new ChoJSONRecordConfiguration
{ 
    Culture = System.Globalization.CultureInfo.InvariantCulture,
    FlattenNode = true
};

using
(
    var r = new ChoJSONReader(jobj, conf)
        .WithField("TimestampGlobal", jsonPath: "TimestampGlobal", fieldType: typeof(DateTime), fieldName: "TimestampGlobal", isArray: false)
        .WithField("TimestampLocal", jsonPath: "TimestampLocal", fieldType: typeof(DateTime), fieldName: "TimestampLocal", isArray: false)
        .WithField("TransactionId", jsonPath: "TransactionId", fieldType: typeof(string), fieldName: "TransactionId", isArray: false)
        .WithField("ComponentId", jsonPath: "ComponentId", fieldType: typeof(string), fieldName: "ComponentId", isArray: false)
        .WithField("TargetComponentId", jsonPath: "TargetComponentId", fieldType: typeof(string), fieldName: "TargetComponentId", isArray: false)
        .WithField("ActionId", jsonPath: "ActionId", fieldType: typeof(string), fieldName: "ActionId", isArray: false)
        .WithField("Duration", jsonPath: "Duration", fieldType: typeof(int), fieldName: "Duration", isArray: false)
        .WithField("MaxTemperature", jsonPath: "AxisData[*].MaxTemperature", fieldType: typeof(decimal), fieldName: "AxisMaxTemperature", isArray: false)
        .WithField("PositionFrom", jsonPath: "AxisData[*].PositionFrom", fieldType: typeof(decimal), fieldName: "AxisPositionFrom", isArray: false)
        .WithField("PositionTo", jsonPath: "AxisData[*].PositionTo", fieldType: typeof(decimal), fieldName: "AxisPositionTo", isArray: false)
        .WithField("Name", jsonPath: "AxisData[*].Name", fieldType: typeof(string), fieldName: "AxisName", isArray: false)
        .WithField("ComponentId", jsonPath: "AxisData[*].ComponentId", fieldType: typeof(string), fieldName: "AxisComponentId", isArray: false)
        .WithField("MaxCurrent", jsonPath: "AxisData[*].MaxCurrent", fieldType: typeof(decimal), fieldName: "AxisMaxCurrent", isArray: false)
    .ErrorMode(ChoErrorMode.ReportAndContinue)
)
{
    using (var w = new ChoCSVWriter(csv)
        .WithDelimiter(",")
        .WithFirstLineHeader())
    {
        w.Write(r);
    }

    System.Console.WriteLine(csv.ToString());
}
```

Thanks!

**Comment by Cinchoo** [comment]
Pls post expected csv file?

**Comment by Cinchoo** [comment]
Released v1.2.1.43-beta4 with fix.

Drafted sample code for review. Take a look and let me know.

https://dotnetfiddle.net/HFLsnH

**Comment by Cinchoo** [comment]
> Another issue - I noticed that specifying any arguments other than the property name `.WithField("MaxCurrent")` for **array** elements causes the csv columns to be empty.
> 
> For example, specifying this `.WithField("MaxCurrent", jsonPath: "$.AxisData[*].MaxCurrent", fieldType: typeof(decimal), fieldName: "AxisMaxCurrent", isArray: true)` results in an empty column.
> 
> Even just trying to rename the field with `.WithField("MaxCurrent", fieldName: "AxisMaxCurrent")` results in an empty column.

After flattening (`FlattenNode = true`) the nodes, JSON turns into array of nodes, each will looks like below

```
{
	"Message": "MsgName",
	"TimestampLocal": "2022-02-02T12:06:18.3400276+11:00",
	"TimestampGlobal": "2022-02-02T00:58:46.1036028Z",
	"TargetComponentId": "compid_a",
	"ActionId": "87990",
	"ComponentId": "compid_b",
	"Duration": 0,
	"TransactionId": "0b10e099-69b8-4a8e-b704-d087ef0d6915"	
	"AxisDataMaxTemperature": 33.1,
	"AxisDataPositionFrom": 2660.0,
	"AxisDataPositionTo": 1311.0,
	"AxisDataName": "Xaxis",
	"AxisDataComponentId": "Xaxis_compid",
	"AxisDataMaxCurrent": 7692.0
}
```
Hence the selection of fields should consider this node structure in mind to specify the jsonpath.

`.WithField("AxisDataMaxCurrent", jsonPath: "$.AxisDataMaxCurrent", fieldType: typeof(decimal), fieldName: "AxisMaxCurrent", isArray: false)`

Hope it helps.

**Comment by gjvanvuuren** [comment]
Thanks for the quick support and the bugfix. The fix in v1.2.1.43-beta4 works for me.

It seems I misunderstood the context of the jsonpath argument. I assumed it describes the structure of the json source. Thanks for clarifying that it's actually the flattened target that's addressed.

What also confused me was the convention: Flattened objects retain their implicit structure, whereas arrays are flattened. So in the following json:

`{
    "Message": "MsgName",
    "TimestampGlobal": "2022-01-01T00:01:12.9461035Z",
    "MachineId": "123456",
    "MachineType": "TypeName",
    "PCName": "pcxyz",
    "Software": {
      "SWVersion": "1.0.0.0",
      "Release": "xyz"
    },
    "AxisData": [
      {
        "Name": "Xaxis",
        "ComponentId": "comp1",
        "MaxCurrent": 12.0
      }
]
}`

the object **SWVersion** has the jsonPath `$.Software.SWVersion`, whereas for the array element **ComponentId** the jsonPath has been compressed to `$.AxisDataComponentId`. This is just a distinction users have to be careful to observe.

Once again - thanks for the very useful library.
